### ISSUE TITLE
TypeScript `createTypeReference` crashes with `TypeError: Cannot read properties of undefined (reading 'get')`

### ISSUE THREAD

**Issue by SukkaW** [microsoft/TypeScript#50254]
# Bug Report

### 🔎 Search Terms

- TypeError: Cannot read properties of undefined (reading 'get')
- createTypeReference
- getJsxPropsTypeFromClassType
- `@compiled/react`

### 🕗 Version & Regression Information

- This is a crash
- I have tested from `@4.4` to `@latest` to `@next` and the crash persists

### 💻 Code

```jsonc
// tsconfig.json
{
  "compilerOptions": {
    "strict": true,
    "module": "esnext",
    "noEmit": true,
    "esModuleInterop": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "react-jsx",
    "jsxImportSource": "@compiled/react" // There goes the culpirit
  },
  "include": [
    "**/*.ts",
    "**/*.tsx"
  ],
  "exclude": [
    "node_modules"
  ]
}

```

```ts
import React from "react";
import { createRoot } from "react-dom/client";
import { ErrorBoundary } from "react-error-boundary"

interface Props extends React.PropsWithChildren<{}> {
  fallback: React.ReactNode;
}

// Note every class component will cause typescript to crash.
// In real-world cases, I am writing an error boundary component which has to be a class component.
export class App extends React.Component<Props, { error: null | Error }> {
  constructor(props: Props) {
    super(props);
  }

  render() {
    return this.props.children;
  }
}

const root = createRoot(document.getElementById("root")!);
root.render(<App />);
```

And I have created a minimal reproduction of the issue:

```bash
git clone https://github.com/SukkaW/compiled-react-ts-crash-repro
cd compiled-react-ts-crash-repro
npx --no-install tsc --noEmit --incremental
# tsc crashes.
```

### 🙁 Actual behavior

With `jsxImportSource` set to `@compiled/css`, any React class component will cause the TypeScript to crash. Change `jsxImportSource` back to `react` the crash disappears.

Here is the crash stack (generated with `typescript@4.7.4`):

```
TypeError: Cannot read properties of undefined (reading 'get')
    at createTypeReference (/path/to/node_modules/typescript/lib/tsc.js:49682:46)
    at getJsxPropsTypeFromClassType (/path/to/node_modules/typescript/lib/tsc.js:60905:27)
    at getEffectiveFirstArgumentForJsxSignature (/path/to/node_modules/typescript/lib/tsc.js:60813:19)
    at checkApplicableSignatureForJsxOpeningLikeElement (/path/to/node_modules/typescript/lib/tsc.js:62996:29)
    at getSignatureApplicabilityError (/path/to/node_modules/typescript/lib/tsc.js:63078:22)
    at chooseOverload (/path/to/node_modules/typescript/lib/tsc.js:63523:25)
    at resolveCall (/path/to/node_modules/typescript/lib/tsc.js:63403:26)
    at resolveJsxOpeningLikeElement (/path/to/node_modules/typescript/lib/tsc.js:64063:20)
    at resolveSignature (/path/to/node_modules/typescript/lib/tsc.js:64084:28)
    at getResolvedSignature (/path/to/node_modules/typescript/lib/tsc.js:64095:26)
```

I have also tried `node --inspect-brk` to get the full stack and code frames:

```
16:07:49.947 VM127 tsc.js:49682 Uncaught TypeError: Cannot read properties of undefined (reading 'get')
    at createTypeReference (VM127 tsc.js:49682:46)
    at getJsxPropsTypeFromClassType (VM127 tsc.js:60905:27)
    at getEffectiveFirstArgumentForJsxSignature (VM127 tsc.js:60813:19)
    at checkApplicableSignatureForJsxOpeningLikeElement (VM127 tsc.js:62996:29)
    at getSignatureApplicabilityError (VM127 tsc.js:63078:22)
    at chooseOverload (VM127 tsc.js:63559:25)
    at resolveCall (VM127 tsc.js:63400:26)
    at resolveJsxOpeningLikeElement (VM127 tsc.js:64063:20)
    at resolveSignature (VM127 tsc.js:64084:28)
    at getResolvedSignature (VM127 tsc.js:64095:26)
createTypeReference @ VM127 tsc.js:49682
getJsxPropsTypeFromClassType @ VM127 tsc.js:60905
getEffectiveFirstArgumentForJsxSignature @ VM127 tsc.js:60813
checkApplicableSignatureForJsxOpeningLikeElement @ VM127 tsc.js:62996
getSignatureApplicabilityError @ VM127 tsc.js:63078
chooseOverload @ VM127 tsc.js:63559
resolveCall @ VM127 tsc.js:63400
resolveJsxOpeningLikeElement @ VM127 tsc.js:64063
resolveSignature @ VM127 tsc.js:64084
getResolvedSignature @ VM127 tsc.js:64095
checkJsxOpeningLikeElementOrOpeningFragment @ VM127 tsc.js:61834
checkJsxElementDeferred @ VM127 tsc.js:61413
checkDeferredNode @ VM127 tsc.js:71870
checkDeferredNodes @ VM127 tsc.js:71834
checkSourceFileWorker @ VM127 tsc.js:71913
checkSourceFile @ VM127 tsc.js:71879
checkSourceFileWithEagerDiagnostics @ VM127 tsc.js:71973
getDiagnosticsWorker @ VM127 tsc.js:71981
getDiagnostics @ VM127 tsc.js:71956
getEmitResolver @ VM127 tsc.js:39980
emitWorker @ VM127 tsc.js:96812
(anonymous) @ VM127 tsc.js:96799
runWithCancellationToken @ VM127 tsc.js:96882
emit @ VM127 tsc.js:96799
getFileEmitOutput @ VM127 tsc.js:98639
updateShapeSignature @ VM127 tsc.js:98879
getFilesAffectedBy @ VM127 tsc.js:98847
getNextAffectedFile @ VM127 tsc.js:99223
getSemanticDiagnosticsOfNextAffectedFile @ VM127 tsc.js:99782
getSemanticDiagnostics @ VM127 tsc.js:99809
emitFilesAndReportErrors @ VM127 tsc.js:101595
emitFilesAndReportErrorsAndGetExitStatus @ VM127 tsc.js:101624
performIncrementalCompilation @ VM127 tsc.js:101807
performIncrementalCompilation @ VM127 tsc.js:104551
executeCommandLineWorker @ VM127 tsc.js:104423
executeCommandLine @ VM127 tsc.js:104471
(anonymous) @ VM127 tsc.js:104765
Module._compile @ VM102 loader:1102
Module._extensions..js @ VM102 loader:1159
Module.load @ VM102 loader:981
Module._load @ VM102 loader:822
executeUserEntryPoint @ VM126 run_main:77
(anonymous) @ VM95 run_main_module:17
```

### 🙂 Expected behavior

The TypeScript shouldn't crash.

### Other useful stuff

- The source code of `@compiled/react`: https://github.com/atlassian-labs/compiled/tree/master/packages/react
- The source code of `@compiled/react/jsx-runtime`: https://github.com/atlassian-labs/compiled/tree/master/packages/react/src/jsx
- The dist of `@compiled/react` (and its `jsx-runtime`) on UNPKG: https://unpkg.com/browse/@compiled/react@0.11.0/
- https://github.com/atlassian-labs/compiled/issues/1282

**Comment by Jack-Works** [comment]
https://github.com/microsoft/TypeScript/pull/50257

**Comment by SukkaW** [comment]
@weswigham Hello there.

Although @Jack-Works has helped fix the issue, I notice that the fix is not included in TS 4.8. Is it intended?

**Comment by kidonng** [comment]
Seems like they forgot to add this to milestones.
